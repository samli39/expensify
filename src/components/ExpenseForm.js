import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from'react-dates';



//parent:AddExpensePage,EditExpensePage

class ExpenseForm extends React.Component{
	state={
		description:this.props.expense?this.props.expense.description:"",
		amount:this.props.expense ? (this.props.expense.amount/100).toString():"",
		note:this.props.expense ? this.props.expense.note : "",
		createdAt:this.props.expense?moment(this.props.expense.createdAt): moment(),
		calenderFocused:false,
		error:""
	}
	//form submition
	onSubmit=(e)=>{
		e.preventDefault();
		if(!this.state.description || !this.state.amount){
			this.setState(()=>({error:"please enter description or amount"}))
		}else{
			this.setState(()=>({error:""}));
			this.props.onSubmit({
				description:this.state.description,
				amount:parseFloat(this.state.amount,10)*100,
				createdAt:this.state.createdAt.valueOf(),
				note:this.state.note
			})
		}
	}
	//description
	onDescriptionChange=(e)=>{
		const description=e.target.value;
		this.setState(()=>({description}));
	}

	//Amount
	onAmountChange=(e)=>{
		const amount=e.target.value;
		//!amount refer to delete amount
		//^\d{1,}-start from number
		//{1,}-at least 1 to infinite
		//()?-appear zero or one time
		// \. - .symbol
		//\d{0,2} - happen 0 to 2 times
		( !amount || (/^\d{1,}(\.\d{0,2})?$/).test(amount))
		&&
		this.setState(()=>({amount}));
	}
	
	//date picker
	onDateChange=(createdAt)=>{
		if(createdAt)
			this.setState(()=>({createdAt}));
	}
	
	//focus on date picker
	onFocusChange=({focused})=>{
		this.setState(()=>({calenderFocused:focused}));
	}
	//note
	onNoteChange=(e)=>{
		const note = e.target.value;
		this.setState(()=>({note}));
	}
	
	render(){
		return(
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					
					{/*description*/}
					<input 
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>

					{/*amount*/}
					<input 
						type="text"
						placeholder="Amount"
						onChange={this.onAmountChange}
						value={this.state.amount}
					/>
				{/*	dates
					numberOfMonths =show how many month
					isOutsideRange =allow pick previous month
				*/}
				<SingleDatePicker 
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calenderFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={()=>false}
					showClearDate={true}
				/>


					
					{/* note */}
					<textarea
						placeholder="Add some not for your expense{optional}"
						value = {this.state.note}
						onChange={this.onNoteChange}
					>

					</textarea>
					<button>Add Expense</button>
				</form>
			</div>

		)
	}
}

export default ExpenseForm;