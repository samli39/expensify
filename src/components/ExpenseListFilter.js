import React from 'react';
import {connect} from 'react-redux';
import {setFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from '../action/action_filter';
import {DateRangePicker} from 'react-dates';

//parent:ExpenseDashboardPahe

export class ExpenseListFilter extends React.Component{
	state={
		calendarFocused:null
	}
	
	// input text for filter text
	onTextChange=(e)=>{
		this.props.setFilter(e.target.value);
	}

	//for sortBy
	onSortByChange=(e)=>{
		(e.target.value === "date") ?
		this.props.sortByDate()
		:
		this.props.sortByAmount();
	}

	//DateRanger Picker
	onDateChange=({startDate,endDate})=>{
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}

	onFocusChange=(calenderFocused)=>{
		this.setState(()=>({calenderFocused}))
	}

	render(){
		return (
			<div>
				{/*for filter text*/}
				<input 
					type="text" 
					value={this.props.filters.text}
					onChange={this.onTextChange}
				/>
				
				{/*option for sortBy*/}
				<select
					value={this.props.filters.sortBy}
					onChange={this.onSortByChange}
					>
					<option value="date">Date</option>
					<option value="amount">Amount</option>


				</select>
				
				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDateChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={()=>false}
					showClearDates={true}
				/>
			</div>

		)
	}
}

const mapStateToProps=(state)=>({
	filters:state.filters
})
const mapDispatchToProps=(dispatch)=>({
	setFilter:(text)=>dispatch(setFilter({text})),
	sortByDate:()=>dispatch(sortByDate()),
	sortByAmount:()=>dispatch(sortByAmount()),
	setStartDate:(date)=>dispatch(setStartDate(date)),
	setEndDate:(date)=>dispatch(setEndDate(date))
})

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter);
// export default ExpenseListFilter;

