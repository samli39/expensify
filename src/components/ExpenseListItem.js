import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
//parent:ExpenseList

const ExpenseListItem=({id,description,createdAt,amount})=>(
	<div>
		<p>description: <Link to ={`/edit/${id}`}>{description}</Link></p>
		<p>amount: {amount} </p>
		<p>createdAt: { moment(createdAt).format("MMMM Do,YYYY") }</p>
	</div>
)



export default ExpenseListItem;