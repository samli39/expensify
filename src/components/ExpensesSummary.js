import React from 'react';
import {connect} from 'react-redux';
import expenseSelector from '../selector/expenseSelector';
import numeral from 'numeral';
import expenseTotal from '../selector/expenseTotal';

//parent:ExpenseDashboardPage
export const ExpensesSummary=(props)=>{
	const amount = expenseTotal(props.expenses);
	const expenseWord = (props.expenses.length === 1) ? 'expense' :'expenses';
	return(
		<div>
			<h1>Viewing {props.expenses.length} {expenseWord} totalling {numeral(amount/100).format('$0,0.00')}</h1>
		</div>
	)
}

const mapStateToProps=(state)=>({
	expenses:expenseSelector(state.expenses,state.filters)
});

//other way

// export const ExpensesSummary =({expenseCount,expenseTotal})=>(
// 	<div>
// 	<h1>Viewing {expenseCount} expenses totalling : {expenseTotal}</h1>
// 	</div>
// 	)




// const mapStateToProps=(state)=>{
// 	const expenses=expenseSelector(state.expenses,state.filters);
// 	return{
// 		expenseCount:expenses.length,
// 		expenseTotal:numeral(getExpensesTotal(expenses)).format("$0,0.00")
// 	}
// }

export default connect(mapStateToProps)(ExpensesSummary);