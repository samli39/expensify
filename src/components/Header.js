import React from 'react';
import {NavLink} from 'react-router-dom';

const Header =()=>(
	<header>
		<NavLink to="/" activeClassName="is-actives" exact>Home </NavLink>
		<br />
		<NavLink to="/create" activeClassName="is-actives">create </NavLink>
		<br />
		<NavLink to="/help" activeClassName="is-actives">Help</NavLink>
	</header>
)

export default Header;