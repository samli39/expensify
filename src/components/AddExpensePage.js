import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../action/action_expenses'


export const AddExpensePage=(props)=>{
	const onSubmit=(expense)=>{
		props.addExpense(expense);
		props.history.push("/");
	}


	return(
		<div>
			it is add expense Page
			<ExpenseForm onSubmit={onSubmit}/>
		</div>

	)
}

const mapDispatchToProps=(dispatch)=>({
	addExpense:(expense)=>dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);