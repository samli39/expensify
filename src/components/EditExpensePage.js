import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../action/action_expenses';

export const EditExpensePage=(props)=>{
	const onSubmit=(expense)=>{
		props.editExpense(props.expense.id,expense);
		props.history.push("/");
	}

	const remove=()=>{
		props.removeExpense(props.expense.id);
		props.history.push('/');
	}

	return(
		<div>
			it is Edit expense Page
			<h1>{props.expense.description}</h1>
			<ExpenseForm onSubmit={onSubmit} expense={props.expense}/>
			<button onClick={remove} > remove </button>
		</div>

	)
}

const mapStateToProps=(state,props)=>({
	expense:state.expenses.find((expense)=>{
		return expense.id === props.match.params.id;
	})
})

const mapDispatchToProps=(dispatch)=>({
	editExpense:(id,update)=>dispatch(editExpense(id,update)),
	removeExpense:(id)=>dispatch(removeExpense(id))
})

//or

// const mapDispatchToProps=(dispatch,props)=>({
// 	editExpense:(id,update)=>dispatch(editExpense(id,update)),
// 	removeExpense:()=>dispatch(removeExpense(props.expense.id))
// })
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);