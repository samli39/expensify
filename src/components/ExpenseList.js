import React from 'react';
import {connect} from 'react-redux';
import expenseSelector from '../selector/expenseSelector';
import ExpenseListItem from './ExpenseListItem';

//parent:ExpenseDashboardPage
export const ExpenseList = (props)=>(
	<div>
		<h1>Expense List </h1>
		{
			( props.expenses.length > 0 )?(
				props.expenses.map((e)=>
						<ExpenseListItem key={e.id} {...e} />
					)
				):(
					<p> there is no item </p>
				)

		}
	</div>
)

const mapStateToProps=(state)=>({
	expenses:expenseSelector(state.expenses,state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
