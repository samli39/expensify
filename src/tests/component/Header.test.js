import React from 'react';
import {Header} from '../../components/Header';
import {shallow} from 'enzyme';

//for testing imported function
// jest.mock('../../action/action_auth', ()=>({startLogout: jest.fn()
// 	.mockReturnValue(new Promise((resolve,reject)=> 
// 		resolve()
// ))}))
// import { startLogout } from '../../action/action_auth';


test('should render Header correctly',()=>{
	const wrapper = shallow(<Header />);
	expect(wrapper).toMatchSnapshot();

})

//for testing imported function
// test('should handle logout',()=>{
// 	const wrapper = shallow(<Header />);
// 	wrapper.find("button").simulate('click');
// 	expect(startLogout).toHaveBeenCalled();	
// })

test('should handle logout',()=>{
	const startLogout=jest.fn();
	const wrapper = shallow(<Header startLogout={startLogout}/>);
	wrapper.find('button').simulate("click");
	expect(startLogout).toHaveBeenCalled();
})