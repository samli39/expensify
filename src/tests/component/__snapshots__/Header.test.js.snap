// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <NavLink
    activeClassName="is-actives"
    aria-current="page"
    exact={true}
    to="/"
  >
    Home 
  </NavLink>
  <br />
  <NavLink
    activeClassName="is-actives"
    aria-current="page"
    to="/create"
  >
    create 
  </NavLink>
  <br />
  <NavLink
    activeClassName="is-actives"
    aria-current="page"
    to="/help"
  >
    Help
  </NavLink>
</header>
`;
